ex3/r_tanaka/main.py

```diff
@@ -0,0 +1,171 @@
+import pandas as pd
+import numpy as np
+import matplotlib.pyplot as plt
+
+
+def plot_linear_regression(x_input, y_input, ax):
+    """plot linear regression
+
+    Args:
+        x_input (ndarray): explanatory variable
+        y_input (ndarray): response variable
+        ax (list): list of axes objects
+    """
+    # find the regression coefficients
+    a = ((x_input * y_input).mean() - (x_input.mean() * y_input.mean())) / ((x_input ** 2).mean() - x_input.mean() ** 2)
+    b = -(a * x_input.mean()) + y_input.mean()
+
+    # prepare x from -5 to 5 in increments of 0.1
+    x = np.arange(-5, 5, 0.1)
+    y = a * x + b
+
+    # plot
+    ax.plot(x, y, color='r', label="regressed")
+    ax.legend()
+
+
+def plot_polynomial_regression(x_input, y_input, deg, ax, alpha=1.0):
+    """plot polynomial regression
+
+    Args:
+        x_input (ndarray): explanatory variable
+        y_input (ndarray): response variable
+        deg (int): degree for polynomial regression
+        ax (list): list of axes objecs
+        alpha(float): regularization term
+    """
+    # create array X
+    list = []
+    for x in x_input:
+        tmp = []
+        for j in range(0, deg + 1):
+            tmp.append(x ** j)
+        list.append(tmp)
+    X = np.array(list, dtype=float)
```

Comment:
list を介した操作は計算時間的にも可読性的にも推奨されません。ndarray の機能を使いましょう。

同一形状の ndarray 同士は基本演算子(+ - * / **) で要素ごとの演算ができることはご存じだと思います。形状が異なる場合もブロードキャスト機能（配列形状を自動で合わせる機能）を使えば for なしで演算可能です。
[参考](https://note.nkmk.me/python-numpy-broadcasting/)

今回のケースでは、形状の異なる配列に対する冪の演算(**)が必要です
```
  # x_input.shape==(datasize,)
  # j.shape==(deg+1,) 
  X = x_input**j
  # X.shape==(datasize, deg+1)???
```
愚直にこれを実行すると 「shape が違います」 みたいなエラーが出ますが、入力に不足している次元を補うことでこれを実現できます
```suggestion
  j = np.arange(deg+1)
  
  # x_input[:,np.newaxis].shape==(datasize, 1)
  # j[np.newaxis,:].shape==(1, deg+1)
  
  # (datasize, deg+1) <- (datasize, 1) ** (1, deg+1)
  X = x_input[:,np.newaxis]**j[np.newaxis,:]
```
ブロードキャストの法則上、実は j の方の次元追加は要らないですが、慣れるまではやっとくといいと思います